// @skills17/test-result/lib/Test.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

class Test {
  manualCheck = false;

  constructor(name, maxPoints, successful, required) {
    this.name = name;
    this.maxPoints = maxPoints;
    this.successful = successful;
    this.required = required;
  }

  getName() {
    return this.name;
  }

  getPoints() {
    return this.successful ? this.maxPoints : 0;
  }

  getMaxPoints() {
    return this.maxPoints;
  }

  isSuccessful() {
    return this.successful;
  }

  isRequired() {
    return this.required;
  }

  setManualCheck(manualCheck) {
    this.manualCheck = manualCheck;
  }

  requiresManualCheck() {
    return this.manualCheck;
  }

  toJSON() {
    return {
      name: this.getName(),
      points: this.getPoints(),
      maxPoints: this.getMaxPoints(),
      successful: this.isSuccessful(),
      required: this.isRequired(),
      manualCheck: this.requiresManualCheck()
    };
  }

}

exports.default = Test;
