// @skills17/karma-helpers/lib/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _taskConfig = _interopRequireDefault(require("@skills17/task-config"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = ({
  plugins = [],
  ...overrides
} = {}) => config => {
  const task = new _taskConfig.default();
  task.loadFromFileSync();
  config.set({
    autoWatch: false,
    basePath: task.getProjectRoot(),
    browsers: ['ChromeHeadless'],
    files: [...task.getSource(), ...task.getTests()],
    hostname: !task.getServe().bind || task.getServe().bind === '127.0.0.1' ? 'localhost' : task.getServe().bind,
    listenAddress: task.getServe().bind ?? '127.0.0.1',
    logLevel: config.LOG_ERROR,
    plugins: [...plugins, require('./reporter')],
    // eslint-disable-line global-require
    reporters: ['skills17'],
    singleRun: true,
    ...overrides
  });
  return config;
};

exports.default = _default;
